File: fixtures/sample-repos/playwright/pages/locators/LoginLocators.ts
import type { Page } from '@playwright/test';

// Locator class: every element-targeting expression lives here.
// No assertions. No navigation logic. Just names.
export class LoginLocators {
  constructor(private page: Page) {}

  get emailInput() {
    return this.page.getByLabel('Email');
  }

  get passwordInput() {
    return this.page.getByLabel('Password');
  }

  get signInButton() {
    return this.page.getByRole('button', { name: 'Sign in' });
  }

  get dashboardHeading() {
    return this.page.getByRole('heading', { name: 'Dashboard' });
  }

  get errorMessage() {
    return this.page.getByText('Invalid credentials');
  }
}

File: fixtures/sample-repos/playwright-dotnet/Pages/Locators/LoginLocators.cs
using Microsoft.Playwright;

namespace Sample.PlaywrightDotnet.Pages.Locators;

public class LoginLocators
{
    private readonly IPage _page;

    private LoginLocators(IPage page) => _page = page;

    public static LoginLocators For(IPage page) => new(page);

    public ILocator EmailInput => _page.GetByLabel("Email");

    public ILocator PasswordInput => _page.GetByLabel("Password");

    public ILocator SignInButton => _page.GetByRole(AriaRole.Button, new() { Name = "Sign in" });

    public ILocator DashboardHeading => _page.GetByRole(AriaRole.Heading, new() { Name = "Dashboard" });

    public ILocator ErrorMessage => _page.GetByText("Invalid credentials");
}
